core/shared/src/test/scala/cats/effect/concurrent/RefSpec.scala

```diff
@@ -0,0 +1,191 @@
+      } yield getAndUpdateResult == 0 && getResult == 1
+
+      op must completeAs(true)
+    }
+
+    "update and get successfully" in {
+      val op = for {
+        r <- Ref[IO].of(0)
+        updateAndGetResult <- r.updateAndGet(_ + 1)
+        getResult <- r.get
+      } yield updateAndGetResult == 1 && getResult == 1
+
+      op must completeAs(true)
+    }
+
+    "access successfully" in {
+      val op = for {
+        r <- Ref[IO].of(0)
+        valueAndSetter <- r.access
+        (value, setter) = valueAndSetter
+        success <- setter(value + 1)
+        result <- r.get
+      } yield success && result == 1
+
+      op must completeAs(true)
+    }
+
+    "access - setter should fail if value is modified before setter is called" in {
+      val op = for {
+        r <- Ref[IO].of(0)
+        valueAndSetter <- r.access
+        (value, setter) = valueAndSetter
+        _ <- r.set(5)
+        success <- setter(value + 1)
+        result <- r.get
+      } yield !success && result == 5
+
+      op must completeAs(true)
+    }
+
+    "access - setter should fail if called twice" in {
+      val op = for {
+        r <- Ref[IO].of(0)
+        valueAndSetter <- r.access
+        (value, setter) = valueAndSetter
+        cond1 <- setter(value + 1)
+        _ <- r.set(value)
+        cond2 <- setter(value + 1)
+        result <- r.get
+      } yield cond1 && !cond2 && result == 0
+
+      op must completeAs(true)
+    }
+
+    "access - setter should fail if called twice" in {
+      val op = for {
+        r <- Ref[IO].of(0)
+        valueAndSetter <- r.access
+        (value, setter) = valueAndSetter
+        cond1 <- setter(value + 1)
+        _ <- r.set(value)
+        cond2 <- setter(value + 1)
+        result <- r.get
+      } yield cond1 && !cond2 && result == 0
+
+      op must completeAs(true)
+    }
+
+    "tryUpdate - modification occurs successfully" in {
+      val op = for {
+        r <- Ref[IO].of(0)
+        result <- r.tryUpdate(_ + 1)
+        value <- r.get
+      } yield result && value == 1
+
+      op must completeAs(true)
+    }
+
+    "tryUpdate - should fail to update if modification has occurred" in {
+      val updateRefUnsafely: Ref[IO, Int] => Unit = (ref: Ref[IO, Int]) => unsafeRun(ref.update(_ + 1))
+
+      val op = for {
+        r <- Ref[IO].of(0)
+        result <- r.tryUpdate { currentValue =>
+          updateRefUnsafely(r)
+          currentValue + 1
+        }
+      } yield result
+
+      op must completeAs(false)
+    }
+
+    "tryModifyState - modification occurs successfully" in {
+      val op = for {
+        r <- Ref[IO].of(0)
+        result <- r.tryModifyState(State.pure(1))
+      } yield result.contains(1)
+
+      op must completeAs(true)
+    }
+
+    "modifyState - modification occurs successfully" in {
+      val op = for {
+        r <- Ref[IO].of(0)
+        result <- r.modifyState(State.pure(1))
+      } yield result == 1
+
+      op must completeAs(true)
+    }
+  }
+
+
+
+
```

Comment:
So


much


blank


space


!

Reply:
😂 I thought the formatter would get rid of this. Will fix!